libtransact/src/scheduler/multi/mod.rs

```diff
@@ -0,0 +1,543 @@
+                    })
+                    .unwrap();
+            });
+
+        let shared_lock = Arc::new(Mutex::new(shared::MultiSchedulerShared::new(schedulers)));
+
+        let core_handle = core::MultiSchedulerCore::new(shared_lock.clone(), core_rx).start();
+
+        MultiScheduler {
+            shared_lock,
+            core_handle: Some(core_handle),
+            core_tx,
+        }
+    }
+
+    pub fn shutdown(mut self) -> std::thread::Result<()> {
+        match self.core_tx.send(core::MultiSchedulerCoreMessage::Shutdown) {
+            Ok(_) => {
+                if let Some(join_handle) = self.core_handle.take() {
+                    return join_handle.join();
+                }
+            }
+            Err(err) => {
+                error!(
+                    "failed to send to MultiScheduler thread during drop: {}",
+                    err
+                );
+            }
+        }
+        Ok(())
+    }
+}
+
+impl Scheduler for MultiScheduler {
+    fn set_result_callback(&mut self, callback: Box<Fn(Option<BatchExecutionResult>) + Send>) {
+        let mut shared = self
+            .shared_lock
+            .lock()
+            .expect("scheduler shared lock is poisoned");
+        shared.set_result_callback(callback);
+    }
+
+    fn add_batch(&mut self, batch: BatchPair) -> Result<(), SchedulerError> {
+        let mut shared = self
+            .shared_lock
+            .lock()
+            .expect("scheduler shared lock is poisoned");
+        if shared.finalized() {
+            return Err(SchedulerError::SchedulerFinalized);
+        }
+        if shared.batch_already_pending(&batch) {
+            return Err(SchedulerError::DuplicateBatch(batch));
+        }
+        shared.add_batch(batch);
+        Ok(())
+    }
+
+    fn cancel(&mut self) -> Vec<BatchPair> {
+        let mut shared = self
+            .shared_lock
+            .lock()
+            .expect("scheduler shared lock is poisoned");
+        shared.cancel()
+    }
+
+    fn finalize(&mut self) {
+        let mut shared = self
+            .shared_lock
+            .lock()
+            .expect("scheduler shared lock is poisoned");
+        shared.finalize();
+    }
+
+    fn take_task_iterator(&mut self) -> Box<dyn Iterator<Item = ExecutionTask> + Send> {
+        // The MultiScheduler passes its sub-schedulers' task iterators directly to the
+        // SubSchedulerHandler
+        unimplemented!("MultiScheduler's take_task_iterator method should not be called");
+    }
+
+    fn new_notifier(&mut self) -> Box<dyn ExecutionTaskCompletionNotifier> {
+        // The MultiScheduler passes its sub-schedulers' notifiers directly to the
+        // SubSchedulerHandler
+        unimplemented!("MultiScheduler's new_notifier method should not be called");
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::scheduler::serial::tests::MockContextLifecycle;
```

Comment:
I don't think the multi-scheduler should depend upon serial scheduler tests. Maybe pop this MockContextLifecycle up a level to the shceduler::tests module?

Reply:
Fixed